README.md
# Instructions

To embed the improved chat server to an edX course, copy and paste the following code to an HTML X-Block element (as raw HTML, **not** text):

~~~html
<link rel="stylesheet" href="https://rawgithub.com/AgileVentures/EduChat/master/app/assets/app.css">
<script type="text/javascript" src="https://rawgithub.com/AgileVentures/EduChat/master/app/chatframe.js"></script>

<div id="chat-div"></div>
~~~

**WARNING**: Embedding this element causes conflicts with the editor, and will be difficult to edit/remove once added. It is recommended to keep a backup of the page somewhere when adding this element.


app/CHANGELOG

changelog

  2ab96158cf 2014-05-12 (MrJ):
    renamed CSS class from 'minimized' to 'enabled' (opposite semantics)
    removed faux clicks and non-functional pre-ready-state DOM checking
    moved premature iframe loading into DOM-ready callback (this was the main bug)
    changed the default state back to disabled (to address legitimate privacy concerns)
    chat is truly disabled by default and is not simply hidden as previous versions
    added "opt-in" button to trigger the actual chat frame loading and login
    changed README.md to instruct loading './app/chatframe.js' to avoid updating xBlock
    much refactoring , reformatting , and a few minor CSS tweaks

  6eccf0f2ac 2014-05-15 (MrJ):
    replaced onload callback to init() with DOM checking (fixes the tab swtching focus bug)
    removed obsolete "review alert" div , function , and interval
    simplified username scraping
    minor bug fix , refactoring , and css tweaks


app/edxframe_by.js
$(function(e) {
  e(window.jQuery, window, document);
}(function($, window, document) {
  
  $(function() {
    var chatBtn = $("#chatframe");
    chatBtn.on("click", function() {
      chatframeContent =
      ['<p><span id="loadingchat">Loading chat...</span>',
       '<iframe id="chatiframe"',
       '       style="position: absolute; top: -9999em; visibility: hidden;"',
//       '       onload="showDelayed(5,this,\'loadingchat\');"',
       '       src="javascript:void(0)"',
       '       height="300" width="100%"></iframe></p>',
       '<link rel="stylesheet" href="https://code.jquery.com/ui/1.10.3/themes/smoothness/jquery-ui.css" />',
       '<style>.no-close .ui-dialog-titlebar-close { display: none; }</style>'
      ].join('\n');

      chatBtn.hide();
      chatBtn.html(chatframeContent);
      $('iframe#chatiframe').on('load', function() {
        showDelayed(5, this, 'loadingchat');
      });

      init();
    });
  });
  
  var chatChannels = "#cs1692x";
  var urlPrefix = 'https://cs1692x.moocforums.org/chatlog/edxframe/CS169.2x/3T2013/';

  var onChatTab = false;
  if (document.title.indexOf("Chat") != -1) {
    onChatTab = true;
  }

  function getContentInContainer(matchClass) {
      var elems = document.getElementsByTagName('*'), i;
      for (i in elems) {
          if((' ' + elems[i].className + ' ').indexOf(' ' + matchClass + ' ')
                  + -1) {
              return elems[i].textContent;
          }
      }
  }

  function getUsername() {
    var name = getContentInContainer("user-link").replace("Dashboard for:", "").replace(/^\s+|\s+$/g, '');
    if (!isNaN(name[0])) {
       name = "_" + name;
    }
    return name;
  }

  function updateChatUrl() {
    username = getUsername();
    document.getElementById('chatiframe').src =
        'https://cs1692x.moocforums.org:9001/' +
        '?channels=' + encodeURIComponent(chatChannels) +
        '&nick=' + encodeURI(username);
  }

  function updateReviewMessage() {
    // Based on http://www.jquery4u.com/json/jsonp-examples/
    $.ajax({
        url: urlPrefix + 'getdate/',
        dataType: 'jsonp',
        type: 'GET',
        async: false,
        contentType: "application/json",
        jsonpCallback: 'jsonCallback',
        success: function (json) {
      if (json.year == 2013 &&
          json.month == 12 &&
          json.day >= 12 && json.day <= 15 &&
          (json.day > 12 || json.hour >= 12) &&
          (json.day < 15 || json.hour < 12) &&
          json.minute >= 50)
      {
          $("#reviewstarting").show();
      } else {
          $("#reviewstarting").hide();
      }
        }
    });
  }

  function init() {
    updateChatUrl();

    if (onChatTab) {
        document.getElementById('chatiframe').height = 600;
    }
    $("#chatframe").show();

    username = getUsername();

    messageContent =
    ['<div class="chapter" id="reviewstarting" style="display:none;">',
     '<h3 aria-label="" class="ui-accordion-header ui-helper-reset ui-state-default ui-corner-all ui-accordion-icons" role="tab" aria-selected="false" tabindex="-1">',
     '<b>The next <a href="https://courses.edx.org/courses/BerkeleyX/CS.169.2x/3T2013/courseware/5208a178b37d45fba5b592daa35f2c14/7a923a96615542cfbff4a5208abbc33d/" style="color:rgb(29, 157, 217);">Interactive Review session</a> is starting in less than 10 minutes. Please arrive in advance in order to complete the consent form.</b>',
     '</h3>',
     '</div><p><p>'
    ].join('\n');

    if ($('#reviewstarting').length == 0) {
      $(".sequence-nav").after(messageContent);
    }

    updateReviewMessage();
    timer = setInterval(function() {
        updateReviewMessage();
    }, 30 * 1000);
  }
  
  function showDelayed(delaySeconds, element, elementHide) {
    setTimeout(function(){
      element.style.position='static';
      element.style.visibility='visible';
      document.getElementById(elementHide).style.display='none';
      $("#chatframe").removeClass('minimized');
    }, 1000*delaySeconds);
  }
}));


app/edxframe_toggleBtn.js
$(function(e) {
  e(window.jQuery, window, document);
}(function($, window, document) {
  
  $(function() {
    var chatBtn = $("#chat_button");
    chatBtn.one("click", function() {
	  chatBtn.html("Loading");
      chatframeContent =
      ['<p><span id="loadingchat">Loading chat...</span>',
       '<iframe id="chatiframe"',
       '       style="position: absolute; top: -9999em; visibility: hidden;"',
//       '       onload="showDelayed(5,this,\'loadingchat\');"',
       '       src="javascript:void(0)"',
       '       height="300" width="100%"></iframe></p>',
       '<link rel="stylesheet" href="https://code.jquery.com/ui/1.10.3/themes/smoothness/jquery-ui.css" />',
       '<style>.no-close .ui-dialog-titlebar-close { display: none; }</style>'
      ].join('\n');

      //chatBtn.hide();
      //chatBtn.html(chatframeContent);
      $('iframe#chatiframe').on('load', function() {
        showDelayed(5, this, 'loadingchat');
      });

      init();
    });
  });
  
  var chatChannels = "#cs1692x";
  var urlPrefix = 'https://cs1692x.moocforums.org/chatlog/edxframe/CS169.2x/3T2013/';

  var onChatTab = false;
  if (document.title.indexOf("Chat") != -1) {
    onChatTab = true;
  }

  function getContentInContainer(matchClass) {
      var elems = document.getElementsByTagName('*'), i;
      for (i in elems) {
          if((' ' + elems[i].className + ' ').indexOf(' ' + matchClass + ' ')
                  + -1) {
              return elems[i].textContent;
          }
      }
  }

  function getUsername() {
    var name = getContentInContainer("user-link").replace("Dashboard for:", "").replace(/^\s+|\s+$/g, '');
    if (!isNaN(name[0])) {
       name = "_" + name;
    }
    return name;
  }

  function updateChatUrl() {
    username = getUsername();
    document.getElementById('chatiframe').src =
        'https://cs1692x.moocforums.org:9001/' +
        '?channels=' + encodeURIComponent(chatChannels) +
        '&nick=' + encodeURI(username);
  }

  function updateReviewMessage() {
    // Based on http://www.jquery4u.com/json/jsonp-examples/
    $.ajax({
        url: urlPrefix + 'getdate/',
        dataType: 'jsonp',
        type: 'GET',
        async: false,
        contentType: "application/json",
        jsonpCallback: 'jsonCallback',
        success: function (json) {
      if (json.year == 2013 &&
          json.month == 12 &&
          json.day >= 12 && json.day <= 15 &&
          (json.day > 12 || json.hour >= 12) &&
          (json.day < 15 || json.hour < 12) &&
          json.minute >= 50)
      {
          $("#reviewstarting").show();
      } else {
          $("#reviewstarting").hide();
      }
        }
    });
  }

  function init() {
    updateChatUrl();

    if (onChatTab) {
        document.getElementById('chatiframe').height = 600;
    }
    $("#chatframe").show();

    username = getUsername();

    messageContent =
    ['<div class="chapter" id="reviewstarting" style="display:none;">',
     '<h3 aria-label="" class="ui-accordion-header ui-helper-reset ui-state-default ui-corner-all ui-accordion-icons" role="tab" aria-selected="false" tabindex="-1">',
     '<b>The next <a href="https://courses.edx.org/courses/BerkeleyX/CS.169.2x/3T2013/courseware/5208a178b37d45fba5b592daa35f2c14/7a923a96615542cfbff4a5208abbc33d/" style="color:rgb(29, 157, 217);">Interactive Review session</a> is starting in less than 10 minutes. Please arrive in advance in order to complete the consent form.</b>',
     '</h3>',
     '</div><p><p>'
    ].join('\n');

    if ($('#reviewstarting').length == 0) {
      $(".sequence-nav").after(messageContent);
    }

    updateReviewMessage();
    timer = setInterval(function() {
        updateReviewMessage();
    }, 30 * 1000);
  }
  
  function showDelayed(delaySeconds, element, elementHide) {
    setTimeout(function(){
      element.style.position='static';
      element.style.visibility='visible';
      //document.getElementById(elementHide).style.display='none';
      $("#chatframe").removeClass('minimized');
      $("#chat_button").html("Close");
      $('#chat_button').on("click", function() {
       $('#chatframe').toggleClass("minimized");
	   if ($('#chatframe').hasClass("minimized")) {
			$("#chat_button").html("Open");
	   } else {
		    $("#chat_button").html("Close");
	   }
    })}, 1000*delaySeconds);
  }
}));



oldPrototype/edxframe2.js
var chatChannels = "#cs1692x";
var urlPrefix = 'https://cs1692x.moocforums.org/chatlog/edxframe/CS169.2x/3T2013/';

var onChatTab = false;
if (document.title.indexOf("Chat") != -1) {
  onChatTab = true;
}

chatframeContent =
['<p><span id="loadingchat">Loading chat...</span>',
 '<iframe id="chatiframe"',
 '       style="position: absolute; top: -9999em; visibility: hidden;"',
 '       onload="showDelayed(5,this,\'loadingchat\');"',
 '       src="javascript:void(0)"',
 '       height="300" width="100%"></iframe></p>',
 '<link rel="stylesheet" href="https://code.jquery.com/ui/1.10.3/themes/smoothness/jquery-ui.css" />',
 '<style>.no-close .ui-dialog-titlebar-close { display: none; }</style>'
].join('\n');

$("#chatframe").hide();
$("#chatframe").html(chatframeContent);

function getContentInContainer(matchClass) {
    var elems = document.getElementsByTagName('*'), i;
    for (i in elems) {
        if((' ' + elems[i].className + ' ').indexOf(' ' + matchClass + ' ')
                > -1) {
            return elems[i].textContent;
        }
    }
}

function getUsername() {
  var n = getContentInContainer("user-link").replace("Dashboard for:", "").replace(/^\s+|\s+$/g, '');
  if (!isNaN(n[0])) {
     n = "_" + n;
  }
  return n;
}

function updateChatUrl() {
  username = getUsername();
  document.getElementById('chatiframe').src =
      'https://cs1692x.moocforums.org:9001/' +
      '?channels=' + encodeURIComponent(chatChannels) +
      '&nick=' + encodeURI(username);
}

if (document.getElementById('chatiframe')) {
  init();
} else {
  window.onload = init;
}

function updateReviewMessage() {
  // Based on http://www.jquery4u.com/json/jsonp-examples/
  $.ajax({
      url: urlPrefix + 'getdate/',
      dataType: 'jsonp',
      type: 'GET',
      async: false,
      contentType: "application/json",
      jsonpCallback: 'jsonCallback',
      success: function (json) {
	  if (json.year == 2013 &&
	      json.month == 12 &&
	      json.day >= 12 && json.day <= 15 &&
	      (json.day > 12 || json.hour >= 12) &&
	      (json.day < 15 || json.hour < 12) &&
	      json.minute >= 50)
	  {
	      $("#reviewstarting").show();
	  } else {
	      $("#reviewstarting").hide();
	  }
      }
  });
}

function init() {
  updateChatUrl();

  if (onChatTab) {
      document.getElementById('chatiframe').height = 600;
  }
  $("#chatframe").show();

  username = getUsername();

  messageContent =
  ['<div class="chapter" id="reviewstarting" style="display:none;">',
   '<h3 aria-label="" class="ui-accordion-header ui-helper-reset ui-state-default ui-corner-all ui-accordion-icons" role="tab" aria-selected="false" tabindex="-1">',
   '<b>The next <a href="https://courses.edx.org/courses/BerkeleyX/CS.169.2x/3T2013/courseware/5208a178b37d45fba5b592daa35f2c14/7a923a96615542cfbff4a5208abbc33d/" style="color:rgb(29, 157, 217);">Interactive Review session</a> is starting in less than 10 minutes. Please arrive in advance in order to complete the consent form.</b>',
   '</h3>',
   '</div><p><p>'
  ].join('\n');

  if ($('#reviewstarting').length == 0) {
    $(".sequence-nav").after(messageContent);
  }

  updateReviewMessage();
  timer = setInterval(function() {
      updateReviewMessage();
  }, 30 * 1000);
}

function showDelayed(delaySeconds, element, elementHide) {
  setTimeout(function(){
    element.style.position='static';
    element.style.visibility='visible';
    document.getElementById(elementHide).style.display='none';
  }, 1000*delaySeconds);
}


app/chatframe_toggleBtn.js

/*\ chatframe_toggleBtn.js - edX embedded chat room load script
|*|
|*|  this version features an enable/disable toggle button persisted by cookie storage
|*|  chat is disabled by default
\*/


(function()
{
  /* constants */

  var STATIC_CHANNELS    = '' ;      // CSV list - must end in ',' comma if non-empty
  var ALLOW_WORLD        = false ;   // useful for offsite development
  var ANONYMOUS_USERNAME = "Anon_" ; // if ALLOW_WORLD
  var CHAT_SERVER_URL    = 'https://cs1692x.moocforums.org:9001/' ;
  var DOM_WAIT_IVL       = 1000 ;
  var IFRAME_WAIT_IVL    = 5000

  var CHAT_STATE_DISABLED              = 'disabled' ;
  var CHAT_STATE_ENABLED               = 'enabled' ;
  var CHAT_STATE_MINIMIZED             = 'minimized' ;
  var CHAT_BTN_TEXTS                   = {} ;
  CHATBTN_TEXT_DISABLED                = "Join Chat" ;
  CHATBTN_TEXT_ENABLED                 = "Hide Chat" ;
  CHATBTN_TEXT_MINIMIZED               = "Show Chat" ;
  CHAT_BTN_TEXTS[CHAT_STATE_DISABLED]  = CHATBTN_TEXT_DISABLED ;
  CHAT_BTN_TEXTS[CHAT_STATE_ENABLED]   = CHATBTN_TEXT_ENABLED ;
  CHAT_BTN_TEXTS[CHAT_STATE_MINIMIZED] = CHATBTN_TEXT_MINIMIZED ;
  var CHAT_CONTAINER_DIV_ID = 'chat-div' ;
  var CHAT_IFRAME_ID        = 'chat-iframe' ;
  var CHAT_LOADBTN_ID       = 'chat-load-button' ;
  var CHAT_TOGGLEBTN_ID     = 'chat-toggle-button' ;
  var DASHBOARD_BTN_CLASS   = 'user-link' ;
  var CHAT_TAB_TITLE        = 'Chat' ;
  var COOKIE_NAME           = 'chat_default' ;
  var CHATBTN_HTML   =
    '<a href="javascript:void(0);" id="' + CHAT_LOADBTN_ID + '">' + CHATBTN_TEXT_DISABLED + '</a>' ;
  var CHATFRAME_HTML =
  [
    '<a href="javascript:void(0);" id="' + CHAT_TOGGLEBTN_ID + '"></a>',
    '<iframe src="javascript:void(0)" id="' + CHAT_IFRAME_ID + '"></iframe>',
    '<link rel="stylesheet" href="https://code.jquery.com/ui/1.10.3/themes/smoothness/jquery-ui.css" />',
    '<style>.no-close .ui-dialog-titlebar-close { display: none; }</style>'
  ].join('\n') ;


  /* functions */

  function init()
  {
    document.getElementById(CHAT_CONTAINER_DIV_ID).style.display = 'block' ;
    if (IsChatEnabled) loadChat() ;
    else
    {
      setChatState(CHAT_STATE_DISABLED) ;
      $('#' + CHAT_CONTAINER_DIV_ID).html(CHATBTN_HTML) ;
      $('#' + CHAT_LOADBTN_ID).on("click" , loadChat) ;
    }
  }

  function loadChat()
  {
    var username = getUsername() ; if (!username && !ALLOW_WORLD) return ;

    $('#' + CHAT_CONTAINER_DIV_ID).html(CHATFRAME_HTML) ;
    $('#' + CHAT_CONTAINER_DIV_ID).display = 'none' ;
    $('#' + CHAT_TOGGLEBTN_ID).html("Loading ...") ;
    document.getElementById(CHAT_IFRAME_ID).src =
      CHAT_SERVER_URL +
      '?channels='    + encodeURIComponent(ChatChannels) +
      '&nick='        + encodeURI(username)              ;

    $('#' + CHAT_IFRAME_ID).on('load' , function()
    {
      if (IsChatTab) document.getElementById(CHAT_IFRAME_ID).height = 600 ;

      // TODO: this should really be an iframe callback
      //       as there is no guarantee that the chat actually loaded
      window.setTimeout(function()
      {
        setChatState(CHAT_STATE_ENABLED) ;
        $('#' + CHAT_CONTAINER_DIV_ID).toggleClass(CHAT_STATE_ENABLED) ;
        $('#' + CHAT_TOGGLEBTN_ID).on("click" , function ()
        {
          $('#' + CHAT_CONTAINER_DIV_ID).toggleClass(CHAT_STATE_ENABLED) ;
          var isEnabled = ($('#' + CHAT_CONTAINER_DIV_ID).hasClass(CHAT_STATE_ENABLED)) ;
          setChatState((isEnabled)? CHAT_STATE_ENABLED : CHAT_STATE_MINIMIZED) ;
        }) ;
      } , IFRAME_WAIT_IVL) ;
    }) ;
  }

  function setChatState(chatState)
  {
    $("#" + CHAT_TOGGLEBTN_ID).html(CHAT_BTN_TEXTS[chatState]) ;
    setStoredState(chatState) ;
  }

  function getStoredState() { return CookieObj.cookie(COOKIE_NAME) ; }

  function setStoredState(chatState)
  {
    var nextState = (chatState == CHAT_STATE_ENABLED)? chatState : CHAT_STATE_DISABLED ;
    CookieObj.cookie(COOKIE_NAME , nextState , { expires: 7 , path: '/' }) ;
  }

  function getCookie($) // TODO: this could be much cleaner if using LocalStorage instead
  {
    var pluses = /\+/g;

    function encode(s) { return config.raw ? s : encodeURIComponent(s) ; }

    function decode(s) { return config.raw ? s : decodeURIComponent(s) ; }

    function stringifyCookieValue(value)
    {
      return encode(config.json ? JSON.stringify(value) : String(value)) ;
    }

    function parseCookieValue(s)
    {
      if (s.indexOf('"') === 0)
        // This is a quoted cookie as according to RFC2068, unescape...
        s = s.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');

      try
      {
        // Replace server-side written pluses with spaces.
        // If we can't decode the cookie, ignore it, it's unusable.
        // If we can't parse the cookie, ignore it, it's unusable.
        s = decodeURIComponent(s.replace(pluses, ' '));
        return config.json ? JSON.parse(s) : s;
      }
      catch(e) {}
    }

    function read(s, converter)
    {
      var value = config.raw ? s : parseCookieValue(s);
      return $.isFunction(converter) ? converter(value) : value;
    }

    var cookieObj = {} ;
    var config = cookieObj.cookie = function (key, value, options)
    {
      // Write
      if (value !== undefined && !$.isFunction(value))
      {
        options = $.extend({}, config.defaults, options);

        if (typeof options.expires === 'number')
        {
          var days = options.expires, t = options.expires = new Date();
          t.setTime(+t + days * 864e+5);
        }

        return (document.cookie =
        [
          encode(key), '=', stringifyCookieValue(value),
          options.expires ? '; expires=' + options.expires.toUTCString() : '', // use expires attribute, max-age is not supported by IE
          options.path    ? '; path=' + options.path : '',
          options.domain  ? '; domain=' + options.domain : '',
          options.secure  ? '; secure' : ''
        ].join(''));
      }

      // Read

      var result = key ? undefined : {};

      // To prevent the for loop in the first place assign an empty array
      // in case there are no cookies at all. Also prevents odd result when
      // calling $.cookie().
      var cookies = document.cookie ? document.cookie.split('; ') : [];

      for (var i = 0, l = cookies.length; i < l; i++)
      {
        var parts = cookies[i].split('=');
        var name = decode(parts.shift());
        var cookie = parts.join('=');

        if (key && key === name)
        {
          // If second argument (value) is a function it's a converter...
          result = read(cookie, value);
          break;
        }

        // Prevent storing a cookie that we couldn't decode.
        if (!key && (cookie = read(cookie)) !== undefined) result[name] = cookie;
      }

      return result;
    };

    config.defaults = {};

    cookieObj.removeCookie = function (key, options)
    {
      if (cookieObj.cookie(key) === undefined) return false;

      // Must not alter options, thus extending a fresh object...
      cookieObj.cookie(key, '', $.extend({}, options, { expires: -1 }));
      return !cookieObj.cookie(key);
    };

    return cookieObj ;
  }

  function getUsername()
  {
    var anonymousUsername = (ANONYMOUS_USERNAME + (new Date()).getTime()) ;
    var dashboardA        = document.getElementsByClassName(DASHBOARD_BTN_CLASS)[0] ;
    if (!dashboardA) return (ALLOW_WORLD)? anonymousUsername : '' ;

    var username = dashboardA.textContent.replace("Dashboard for:" , "").replace(/^\s+|\s+$/g , '') ;
    if (!isNaN(username[0])) username = "_" + username ;

    return username ;
  }


/* main */

  var CookieObj     = getCookie(jQuery) ; // TODO: use LocalStorage
  var ChatChannels  = STATIC_CHANNELS + "#" + window.location.pathname.split('/')[3] ;
  var IsChatTab     = (~document.title.indexOf(CHAT_TAB_TITLE)) ;
  var IsChatEnabled = (getStoredState() === CHAT_STATE_ENABLED) ;

  DomWaitIvl = window.setInterval(function()
  {
    if (!document.getElementById(CHAT_CONTAINER_DIV_ID)) return ;

    window.clearInterval(DomWaitIvl) ; init() ;
  } , DOM_WAIT_IVL) ;

})();


app/chatframe.js

/*\ chatframe_toggleBtn.js - edX embedded chat room load script
|*|
|*|  this version features an enable/disable toggle button persisted by cookie storage
|*|  chat is disabled by default
\*/


(function()
{
  /* constants */

  var CHAT_SERVER_URL    = 'https://cs1692x.moocforums.org:9001/' ;
  var STATIC_CHANNELS    = '' ;      // CSV list - must end in ',' comma if non-empty
  var ALLOW_WORLD        = false ;   // useful for offsite development
  var ANONYMOUS_USERNAME = "Anon_" ; // if ALLOW_WORLD
  var DOM_WAIT_IVL       = 1000 ;
  var IFRAME_WAIT_IVL    = 5000 ;

  var CHAT_STATE_DISABLED              = 'disabled' ;
  var CHAT_STATE_ENABLED               = 'enabled' ;
  var CHAT_STATE_MINIMIZED             = 'minimized' ;
  var CHAT_BTN_TEXTS                   = {} ;
  CHATBTN_TEXT_DISABLED                = "Join Chat" ;
  CHATBTN_TEXT_ENABLED                 = "Hide Chat" ;
  CHATBTN_TEXT_MINIMIZED               = "Show Chat" ;
  CHAT_BTN_TEXTS[CHAT_STATE_DISABLED]  = CHATBTN_TEXT_DISABLED ;
  CHAT_BTN_TEXTS[CHAT_STATE_ENABLED]   = CHATBTN_TEXT_ENABLED ;
  CHAT_BTN_TEXTS[CHAT_STATE_MINIMIZED] = CHATBTN_TEXT_MINIMIZED ;
  var CHAT_CONTAINER_DIV_ID = 'chat-div' ;
  var CHAT_IFRAME_ID        = 'chat-iframe' ;
  var CHAT_LOADBTN_ID       = 'chat-load-button' ;
  var CHAT_TOGGLEBTN_ID     = 'chat-toggle-button' ;
  var DASHBOARD_BTN_CLASS   = 'user-link' ;
  var CHAT_TAB_TITLE        = 'Chat' ;
  var COOKIE_NAME           = 'chat_default' ;
  var CHATBTN_HTML   =
    '<a href="javascript:void(0);" id="' + CHAT_LOADBTN_ID + '">' + CHATBTN_TEXT_DISABLED + '</a>' ;
  var CHATFRAME_HTML =
  [
    '<a href="javascript:void(0);" id="' + CHAT_TOGGLEBTN_ID + '"></a>',
    '<iframe src="javascript:void(0)" id="' + CHAT_IFRAME_ID + '"></iframe>',
    '<link rel="stylesheet" href="https://code.jquery.com/ui/1.10.3/themes/smoothness/jquery-ui.css" />',
    '<style>.no-close .ui-dialog-titlebar-close { display: none; }</style>'
  ].join('\n') ;


  /* functions */

  function init()
  {
    document.getElementById(CHAT_CONTAINER_DIV_ID).style.display = 'block' ;
    if (IsChatEnabled) loadChat() ;
    else
    {
      setChatState(CHAT_STATE_DISABLED) ;
      $('#' + CHAT_CONTAINER_DIV_ID).html(CHATBTN_HTML) ;
      $('#' + CHAT_LOADBTN_ID).on("click" , loadChat) ;
    }
  }

  function loadChat()
  {
    var username = getUsername() ; if (!username && !ALLOW_WORLD) return ;

    $('#' + CHAT_CONTAINER_DIV_ID).html(CHATFRAME_HTML) ;
    $('#' + CHAT_CONTAINER_DIV_ID).display = 'none' ;
    $('#' + CHAT_TOGGLEBTN_ID).html("Loading ...") ;
    document.getElementById(CHAT_IFRAME_ID).src =
      CHAT_SERVER_URL +
      '?channels='    + encodeURIComponent(ChatChannels) +
      '&nick='        + encodeURI(username)              ;

    $('#' + CHAT_IFRAME_ID).on('load' , function()
    {
      if (IsChatTab) document.getElementById(CHAT_IFRAME_ID).height = 600 ;

      // TODO: this should really be an iframe callback
      //       as there is no guarantee that the chat actually loaded
      window.setTimeout(function()
      {
        setChatState(CHAT_STATE_ENABLED) ;
        $('#' + CHAT_CONTAINER_DIV_ID).toggleClass(CHAT_STATE_ENABLED) ;
        $('#' + CHAT_TOGGLEBTN_ID).on("click" , function ()
        {
          $('#' + CHAT_CONTAINER_DIV_ID).toggleClass(CHAT_STATE_ENABLED) ;
          var isEnabled = ($('#' + CHAT_CONTAINER_DIV_ID).hasClass(CHAT_STATE_ENABLED)) ;
          setChatState((isEnabled)? CHAT_STATE_ENABLED : CHAT_STATE_MINIMIZED) ;
        }) ;
      } , IFRAME_WAIT_IVL) ;
    }) ;
  }

  function setChatState(chatState)
  {
    $("#" + CHAT_TOGGLEBTN_ID).html(CHAT_BTN_TEXTS[chatState]) ;
    setStoredState(chatState) ;
  }

  function getStoredState() { return CookieObj.cookie(COOKIE_NAME) ; }

  function setStoredState(chatState)
  {
    var nextState = (chatState == CHAT_STATE_ENABLED)? chatState : CHAT_STATE_DISABLED ;
    CookieObj.cookie(COOKIE_NAME , nextState , { expires: 7 , path: '/' }) ;
  }

  function getCookie($) // TODO: this could be much cleaner if using LocalStorage instead
  {
    var pluses = /\+/g;

    function encode(s) { return config.raw ? s : encodeURIComponent(s) ; }

    function decode(s) { return config.raw ? s : decodeURIComponent(s) ; }

    function stringifyCookieValue(value)
    {
      return encode(config.json ? JSON.stringify(value) : String(value)) ;
    }

    function parseCookieValue(s)
    {
      if (s.indexOf('"') === 0)
        // This is a quoted cookie as according to RFC2068, unescape...
        s = s.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');

      try
      {
        // Replace server-side written pluses with spaces.
        // If we can't decode the cookie, ignore it, it's unusable.
        // If we can't parse the cookie, ignore it, it's unusable.
        s = decodeURIComponent(s.replace(pluses, ' '));
        return config.json ? JSON.parse(s) : s;
      }
      catch(e) {}
    }

    function read(s, converter)
    {
      var value = config.raw ? s : parseCookieValue(s);
      return $.isFunction(converter) ? converter(value) : value;
    }

    var cookieObj = {} ;
    var config = cookieObj.cookie = function (key, value, options)
    {
      // Write
      if (value !== undefined && !$.isFunction(value))
      {
        options = $.extend({}, config.defaults, options);

        if (typeof options.expires === 'number')
        {
          var days = options.expires, t = options.expires = new Date();
          t.setTime(+t + days * 864e+5);
        }

        return (document.cookie =
        [
          encode(key), '=', stringifyCookieValue(value),
          options.expires ? '; expires=' + options.expires.toUTCString() : '', // use expires attribute, max-age is not supported by IE
          options.path    ? '; path=' + options.path : '',
          options.domain  ? '; domain=' + options.domain : '',
          options.secure  ? '; secure' : ''
        ].join(''));
      }

      // Read

      var result = key ? undefined : {};

      // To prevent the for loop in the first place assign an empty array
      // in case there are no cookies at all. Also prevents odd result when
      // calling $.cookie().
      var cookies = document.cookie ? document.cookie.split('; ') : [];

      for (var i = 0, l = cookies.length; i < l; i++)
      {
        var parts = cookies[i].split('=');
        var name = decode(parts.shift());
        var cookie = parts.join('=');

        if (key && key === name)
        {
          // If second argument (value) is a function it's a converter...
          result = read(cookie, value);
          break;
        }

        // Prevent storing a cookie that we couldn't decode.
        if (!key && (cookie = read(cookie)) !== undefined) result[name] = cookie;
      }

      return result;
    };

    config.defaults = {};

    cookieObj.removeCookie = function (key, options)
    {
      if (cookieObj.cookie(key) === undefined) return false;

      // Must not alter options, thus extending a fresh object...
      cookieObj.cookie(key, '', $.extend({}, options, { expires: -1 }));
      return !cookieObj.cookie(key);
    };

    return cookieObj ;
  }

  function getUsername()
  {
    var anonymousUsername = (ANONYMOUS_USERNAME + (new Date()).getTime()) ;
    var dashboardA        = document.getElementsByClassName(DASHBOARD_BTN_CLASS)[0] ;
    if (!dashboardA) return (ALLOW_WORLD)? anonymousUsername : '' ;

    var username = dashboardA.textContent.replace("Dashboard for:" , "").replace(/^\s+|\s+$/g , '') ;
    if (!isNaN(username[0])) username = "_" + username ;

    return username ;
  }


/* main */

  var CookieObj     = getCookie(jQuery) ; // TODO: use LocalStorage
  var ChatChannels  = STATIC_CHANNELS + "#" + window.location.pathname.split('/')[3] ;
  var IsChatTab     = (~document.title.indexOf(CHAT_TAB_TITLE)) ;
  var IsChatEnabled = (getStoredState() === CHAT_STATE_ENABLED) ;

  DomWaitIvl = window.setInterval(function()
  {
    if (!document.getElementById(CHAT_CONTAINER_DIV_ID)) return ;

    window.clearInterval(DomWaitIvl) ; init() ;
  } , DOM_WAIT_IVL) ;

})();
